pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,633 @@
+		RetryPeriod:   2 * time.Second,
+		Callbacks:     callbacks,
+		Name:          "leader-election-controller",
+	})
+	return nil, nil
+}
+
+func NewController(leaseInformer coordinationv1.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	klog.Infof("NewController")
+	c := &Controller{
+		leaseInformer:          leaseInformer,
+		leaseCandidateInformer: leaseCandidateInformer,
+		leaseQueue:             workqueue.NewTypedRateLimitingQueueWithConfig(workqueue.DefaultTypedControllerRateLimiter[string](), workqueue.TypedRateLimitingQueueConfig[string]{Name: controllerName}),
+		leaseClient:            leaseClient,
+		leaseCandidateClient:   leaseCandidateClient,
+
+		// TODO: What to do if the size limit is reached?
+		// realistically we'd have a lot less than 1000 leases that all need to be updated at the same time.
+		// Is there any harm to naively spinning a goroutine for each election?
+		// I don't estimate the load on the apiserver to be that much
+		// Need to write producer code to handle? It is unsafe to block on send.
+		electionCh:               make(chan election, 1000),
+		electionCandidateTracker: make(map[leaderLeaseID]*candidateStatus),
+	}
+	leaseSynced, err := leaseInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueLease(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueLease(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.leaseDeleted(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	leaseCandidateSynced, err := leaseCandidateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueCandidate(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueCandidate(newObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	c.leaseSynced = leaseSynced.HasSynced
+	c.leaseCandidateSynced = leaseCandidateSynced.HasSynced
+	return c, nil
+}
+
+func (c *Controller) enqueueLease(obj any) {
+	if lease, ok := obj.(*v1.Lease); ok {
+		// TODO: handle namespaces
+		key, err := controller.KeyFunc(lease)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", lease, err))
+		}
+		c.leaseQueue.Add(key)
+	}
+}
+
+func (c *Controller) enqueueCandidate(obj any) {
+	if lease, ok := obj.(*v1alpha1.LeaseCandidate); ok {
+		// TODO: handle namespaces
+		// key, err := controller.KeyFunc(lease)
+		// if err != nil {
+		// 	utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", lease, err))
+		// }
+		_ = c.reconcileLeaseCandidate(lease)
+		// How to handle invalid lease?
+	}
+}
+
+func (c *Controller) leaseDeleted(oldObj any) {
+	klog.Infof("leaseDeleted")
+	if lease, ok := oldObj.(*v1.Lease); ok {
```

Comment:
you need to handle tombstones, no?

Reply:
https://github.com/kubernetes/kubernetes/blob/06c7d9a565fc7da8ec726e515753da619bc500e6/pkg/proxy/config/config.go#L145-L157

Reply:
Is this still necessary now that we use KeyFunc and enqueue the lease on deletion?